NavCog/Model/StateMachine/NavMachine.m

```diff
@@ -355,9 +369,171 @@ - (void)startNavigationOnTopoMap:(TopoMap *)topoMap fromNodeWithName:(NSString *
+    _speechEnabled = speechEnabled;
+    _clickEnabled = clickEnabled;
+    
+    // search a path
+    _topoMap = topoMap;
+    _pathNodes = nil;
+    if (![fromNodeName isEqualToString:NSLocalizedString(@"currentLocation", @"Current Location")]) {
+        _pathNodes = [_topoMap findShortestPathFromNodeWithName:fromNodeName toNodeWithName:toNodeName];
+        [self initializeWithPathNodes:_pathNodes];
+        _isStartFromCurrentLocation = false;
+        _isNavigationStarted = true;
+        [_delegate navigationReadyToGo];
+    } else {
+        _destNodeName = toNodeName;
+        _isStartFromCurrentLocation = true;
+        _isNavigationStarted = false;
+    }
+    
+    dispatch_queue_t queue = dispatch_queue_create("com.navcog.logsimulatorqueue", NULL);
+    
+    unsigned long int arraySize = [timesArray count];
+    
+    dispatch_async(queue, ^{
+
+        NSDate* time = startTime;
+        
+        for (int i=0; i < arraySize; i++) {
+            
+        
+            if ([objectsArray[i] isKindOfClass: [NSArray class]]) {
+                //create
+                NSTimeInterval waitTime = [timesArray[i] timeIntervalSinceDate:time];
+                
+                NSArray* beacons = objectsArray[i];
+                
+                //call beacons
+                [NSThread sleepForTimeInterval:waitTime];
+                dispatch_sync(dispatch_get_main_queue(), ^{
+                    [self receivedBeaconsArray: beacons];
+                });
+                
+                time = timesArray[i];
+                
+            } else if ([objectsArray[i] isKindOfClass: [NSMutableDictionary class]]) {
+                
+                NSTimeInterval waitTime = [timesArray[i] timeIntervalSinceDate:time];
+                
+                NSMutableDictionary* motionData = objectsArray[i];
+                
+                //call motion
+                    [NSThread sleepForTimeInterval:waitTime];
+                dispatch_sync(dispatch_get_main_queue(), ^{
+                    [self triggerMotionWithData:motionData];
+                });
+                
+                time = timesArray[i];
+                
+            }
+        
+        }
+    });
+    
+}
+
+- (NSMutableArray *)loadLogList {
+    NSFileManager *fm = [NSFileManager defaultManager];
+    
+    NSString* documentsPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) objectAtIndex:0];
+    NSArray* allFileNames = [fm contentsOfDirectoryAtPath:documentsPath error:nil];
+    NSMutableArray* logList = [[NSMutableArray alloc] init];
+    
+    for (NSString* fileName in allFileNames) {
+        if([[fileName lowercaseString] hasSuffix:@".log"]) {
+            [logList addObject:[documentsPath stringByAppendingPathComponent:fileName]];
+        }
+    }
+    
+    return logList;
+}
+
 - (void)stopNavigation {
     [self stopAudio];
-    [_beaconManager stopRangingBeaconsInRegion:_beaconRegion];
+    if (!(_logReplay)) {
```

Comment:
Maybe we write a wrapper around startRanging and stopRanging to handle the `_logReplay` check.